src/metalookup/features/addblock_based.py

```diff
@@ -1,9 +1,127 @@
+from typing import Optional
+from unittest import mock
+
+import adblockparser
+from aiohttp import ClientSession
+
+from metalookup.app.models import Explanation, StarCase
+from metalookup.core.extractor import Extractor, download_tag_lists
+from metalookup.core.website_manager import WebsiteData
+from metalookup.lib.tools import runtime
+
+_FOUND_LIST_MATCHES = "Found list matches"
+_FOUND_NO_LIST_MATCHES = "Found no list matches"
+
+
+class AdBlockBasedExtractor(Extractor[set[str]]):
+    """
+    Base class for features to be extracted with `adblockparser` package.
+
+    The returned extra data corresponds to a sample of the blocked links.
+    The sample size can be configured with the limit member variable.
+    """
+
+    urls: list[str] = []  # needs to be provided by derived class
+
+    def __init__(self, logger: Optional[Logger] = None):
+        self.logger = logger or logging.getLogger(__name__)
+        self.adblock_parser_options = {
+            "script": True,
+            "image": True,
+            "stylesheet": True,
+            "object": True,
+            "xmlhttprequest": True,
+            "object-subrequest": True,
+            "subdocument": True,
+            "document": True,
+            "elemhide": True,
+            "other": True,
+            "background": True,
+            "xbl": True,
+            "ping": True,
+            "dtd": True,
+            "media": True,
+            "third-party": True,
+            "match-case": True,
+            "collapse": True,
+            "donottrack": True,
+            "websocket": True,
+            "domain": "",
+        }
+        self.rules: AdblockRules = None  # noqa will be initialized in async setup call
+        self.limit: Optional[int] = 2
+        """
+        Stop after finding a given amount of to be blocked links.
+        For websites with a lot of referenced content, this can significantly speed up the extraction process,
+        as once it is clear that there will be blocked content, there is no more need to check for more blocked content.
+        If set to none, then all links will be checked.
+        """
+
+    async def setup(self) -> None:
+        """Fetch tag lists from configured urls."""
+        assert self.urls != [], "Missing url specification of tag lists"
+        assert self.rules is None, "rules must be initialized in async setup call"
+
+        # patch the adblockparser python package to work with googles re2 package.
+        # re2 seems to be significantly faster than the builtin re package.
+        # fixme: eventually we may want to abandon usage of the adblockparser package all-together and transition
+        #        to a more modern alternative (e.g. the adblock package which is a python wrapper for a native
+        #        rust implementation that would probably speed things up significantly)
+        import re
+
+        def _combined_regex(regexes, flags=re.IGNORECASE, use_re2=False, max_mem=None):
```

Comment:
Why are you reimplementing this function?

Reply:
because the one shipped with the `adblockparser` package does not work with google-re2 (the google-re2 API is slightly different from what `adblockparser` expects)